client/pdo/client/controller/commands/send.py

```diff
@@ -46,32 +46,28 @@ def send_to_contract(state, save_file, enclave, message, quiet=False, wait=False
         raise Exception('unable to load the contract')
 
     # ---------- set up the enclave service ----------
-    enclave_names = state.get(['Service', 'EnclaveServiceNames'], [])
-    if len(enclave_names) > 0: #use the database to get the list of enclaves for the contract
-        logger.info('Using eservice database to look up service URL for the contract enclave')
-        try:
-            eservice_to_use = random.choice(state.get(['Service', 'EnclaveServiceNames']))
-            enclave_client = db.get_client_by_name(eservice_to_use)
-        except Exception as e:
-            logger.exception('Unable to get the eservice clients using the eservice database: %s', str(e)) 
-            raise Exception from e
-    else:
+    if eservice_url and eservice_url != 'random' :
         try :
-            enclave_client = get_enclave_service(state, enclave)
+            eservice_client = EnclaveServiceClient(eservice_url)
         except Exception as e :
             raise Exception('unable to connect to enclave service; {0}'.format(str(e)))
 
-    try :
-        # this is just a sanity check to make sure the selected enclave
-        # has actually been provisioned
-        contract.get_state_encryption_key(enclave_client.enclave_id)
-    except KeyError as ke :
-        logger.error('selected enclave is not provisioned')
-        sys.exit(-1)
+        if eservice_client.enclave_id not in contract.provisioned_enclaves :
+            raise Exception('requested enclave not provisioned for the contract; %s', eservice_url)
+    else :
+        enclave_id = random.choice(contract.provisioned_enclaves)
+        eservice_info = eservice_db.get_info_by_id(enclave_id)
```

Comment:
shouldn't the preferred enclave be checked here, and used if present?

Reply:
it is.

Reply:
the contract is read in line 44, any preferred enclave set in the extra_data field does not seem to be checked

Reply:
the contract is read in line 44, any preferred enclave set in the extra_data field does not seem to be used